transactions/src/main/java/org/jboss/as/txn/integration/JBossContextXATerminator.java

```diff
@@ -0,0 +1,159 @@
+
+import org.jboss.as.txn.logging.TransactionLogger;
+import org.jboss.tm.JBossXATerminator;
+import org.wildfly.transaction.client.ContextTransactionManager;
+import org.wildfly.transaction.client.ImportResult;
+import org.wildfly.transaction.client.LocalTransaction;
+import org.wildfly.transaction.client.LocalTransactionContext;
+
+import com.arjuna.ats.internal.jta.transaction.arjunacore.jca.SubordinationManager;
+
+public class JBossContextXATerminator implements JBossXATerminator {
+
+    private final LocalTransactionContext localTransactionContext;
+    private final XATerminator contextXATerminator;
+    private final JBossXATerminator jbossXATerminator;
+
+    public JBossContextXATerminator(LocalTransactionContext transactionContext, JBossXATerminator jbossXATerminator) {
+        this.localTransactionContext = transactionContext;
+        this.contextXATerminator = transactionContext.getXATerminator();
+        this.jbossXATerminator = jbossXATerminator;
+    }
+
+    @Override
+    public void commit(Xid xid, boolean onePhase) throws XAException {
+        contextXATerminator.commit(xid, onePhase);
+    }
+
+    @Override
+    public void forget(Xid xid) throws XAException {
+        contextXATerminator.forget(xid);
+    }
+
+    @Override
+    public int prepare(Xid xid) throws XAException {
+        return contextXATerminator.prepare(xid);
+    }
+
+    @Override
+    public Xid[] recover(int flag) throws XAException {
+        return contextXATerminator.recover(flag);
+    }
+
+    @Override
+    public void rollback(Xid xid) throws XAException {
+        contextXATerminator.rollback(xid);
+    }
+
+
+    /**
+     * <p>
+     * Interception of register work call to get transaction being imported to wildfly transacton client.
+     * <p>
+     * For importing a transaction Wildfly transaction client eventually calls {@link SubordinationManager}
+     * as Narayana {@link XATerminator}s do. This wrapping then let wildfly transacton client to register the transaction
+     * for itself, wildfly transacton client then import transaction to Narayana too and finally this method
+     * uses Narayana's {@link XATerminator} to register all {@link Work}s binding.<br>
+     * Narayana's {@link XATerminator} tries to import transaction too but as transaction is already
+     * imported it just gets instance of transaction already imported via call of wildfly transacton client.
+     */
+    @Override
+    public void registerWork(Work work, Xid xid, long timeout) throws WorkCompletedException {
+        try {
+            // jca provides timeout in milliseconds, SubordinationManager expects seconds
+            int timeout_seconds = (int) timeout/1000;
+            // unlimited timeout for jca means -1 which fails in wfly client
+            if(timeout_seconds <= 0) timeout_seconds = ContextTransactionManager.getGlobalDefaultTransactionTimeout();
+            localTransactionContext.findOrImportTransaction(xid, timeout_seconds);
```

Comment:
This and the startWork look like the real heart of this patch. I see no issue with the way you are proposing to change this behaviour and so assuming the rest of my comments are addressed I approve the change